### PR TITLE
feat(utxo-lib): add psbt getSignatureValidationArray function

### PR DESCRIPTION
**Pull request by saravanan7mani**
To check given pub keys are used to sign given PSBT transaction input.

TICKET: BG-58455

### CODE REVIEW

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 1]
```diff
@@ -218,10 +218,18 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint>> extends Psbt {
     return results.reduce((final, res) => res && final, true);
   }
 
-  validateTaprootSignaturesOfInput(inputIndex: number): boolean {
+  validateTaprootSignaturesOfInput(inputIndex: number, pubkey?: Buffer): boolean {
     const input = this.data.inputs[inputIndex];
-    const mySigs = (input || {}).tapScriptSig;
-    if (!input || !mySigs || mySigs.length < 1) throw new Error('No signatures to validate');
+    const tapSigs = (input || {}).tapScriptSig;
+    if (!input || !tapSigs || tapSigs.length < 1) throw new Error('No signatures to validate');
```
Comment (rushilbg) [inline modules/utxo-lib/src/bitgo/UtxoPsbt.ts:224]: Nit: For readability and consistency in the repo.
```suggestion
    if (!input || !tapSigs || tapSigs.length < 1) {
      throw new Error('No signatures to validate');
    }
```

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 2]
```diff
@@ -218,10 +218,18 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint>> extends Psbt {
...
+    let mySigs;
+    if (pubkey) {
+      const xOnlyPubkey = toXOnlyPublicKey(pubkey);
+      mySigs = tapSigs.filter((sig) => sig.pubkey.equals(xOnlyPubkey));
+      if (mySigs.length < 1) throw new Error('No signatures for this pubkey');
```
Comment (rushilbg) [inline modules/utxo-lib/src/bitgo/UtxoPsbt.ts:229]: Same here

**Review by rushilbg — DISMISSED** [review]

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 3]
```diff
@@ -241,6 +253,24 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint>> extends Psbt {
     return results.every((res) => res === true);
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/UtxoPsbt.ts:253]: What's interesting is that `results.every()` returns `true` for empty arrays, which IMO is a bug even without this change. Maybe we should change this to `0 < results.length && results.every`

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 4]
```diff
@@ -241,6 +253,24 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint>> extends Psbt {
...
+    const input = checkForInput(this.data.inputs, inputIndex);
+    const isP2tr = input.tapScriptSig?.length;
+    return publicKeys.map((publicKey) => {
+      try {
+        return isP2tr
+          ? this.validateTaprootSignaturesOfInput(inputIndex, publicKey)
+          : this.validateSignaturesOfInput(inputIndex, (p, m, s) => eccLib.verify(m, p, s), publicKey);
+      } catch (err) {
+        return false;
+      }
```
Comment (OttoAllmendinger) [inline modules/utxo-lib/src/bitgo/UtxoPsbt.ts:270]: Not a fan of this overly broad `catch` here. There are a number of errors that can stem from implementation errors unrelated to sig verification that we want to know about. Without the `throw` in `this.validateTaprootSignaturesOfInput` I believe we can do without the `catch` here.

**Review by OttoAllmendinger — CHANGES_REQUESTED** [review]

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts (hunk 3 above)
Comment (OttoAllmendinger) [reply modules/utxo-lib/src/bitgo/UtxoPsbt.ts:253]: actually to stay compatible with with https://github.com/BitGo/bitcoinjs-lib/blob/master/ts_src/psbt.ts#L435 let's leave it as-is

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts [hunk 5]
```diff
@@ -218,10 +218,22 @@ export class UtxoPsbt<Tx extends UtxoTransaction<bigint>> extends Psbt {
...
+    let mySigs;
+    if (pubkey) {
+      const xOnlyPubkey = toXOnlyPublicKey(pubkey);
+      mySigs = tapSigs.filter((sig) => sig.pubkey.equals(xOnlyPubkey));
+      if (mySigs.length < 1) {
+        throw new Error('No signatures for this pubkey');
+      }
+    } else {
+      mySigs = tapSigs;
+    }
```
[the remark this answers is not in the document]
Comment (OttoAllmendinger) [reply modules/utxo-lib/src/bitgo/UtxoPsbt.ts:236]: actually to be compatible with https://github.com/BitGo/bitcoinjs-lib/blob/master/ts_src/psbt.ts#L435 let's throw instead

File: modules/utxo-lib/src/bitgo/UtxoPsbt.ts (hunk 4 above)
Comment (OttoAllmendinger) [reply modules/utxo-lib/src/bitgo/UtxoPsbt.ts:270]: this comment remains valid - to be compatible with https://github.com/BitGo/bitcoinjs-lib/blob/master/ts_src/psbt.ts#L435 we should remove the `catch`

Comment (saravanan7mani) [reply modules/utxo-lib/src/bitgo/UtxoPsbt.ts]: Errors for unsigned input or input with no matching signature for given public key are caught and returned as false. All other errors, propagating the error.

**Review by OttoAllmendinger — APPROVED** [review]

**Pull request merged**